Hibernate Reactive Gotcha (#97)
lunatech-labs/lunatech-blog: posts/2023-05-19-hibernate-reactive-gotchas.adoc

== How to prevent this?

=== If your own code evaluates Uni's in parallel

If your own code evaluates Uni's in parallel, you have several options:

The most obvious option is to use `.usingConcurrencyOf(1)` when using `Uni.join()` or `Uni.combine()`, for example:

   Uni.join().all(one, two).usingConcurrencyOf(1); // Will evaluate `one` and `two` sequentially.

Another way is two use `flatMap` for more explicit sequential evaluation:

    one.flatMap(x -> two.map(y -> x + " " + y)); // Will evaluate `one` and `two` sequentially.

=== If someone else's code evaluates Uni's in parallel

Sometimes you pass ``Uni``s from Hibernate Reactive to some library for further processing. In that case, you may not have control over how this library evaluates the ``Uni``s.

This bit us in the past, when using Hibernate Reactive together with Smallrye GraphQL: https://github.com/quarkusio/quarkus/issues/32870

> In this scenario, the GraphQL engine decides to evaluate Uni's in parallel (which it _should_, according to the GraphQL spec!), which is not allowed for Hibernate ``Uni``'s.

Reviewer: ```suggestion
In this scenario, the GraphQL engine decides to evaluate ``Uni``s in parallel (which it _should_, according to the GraphQL spec!), which is not allowed for Hibernate ``Uni``s.
```

In the next sections we'll dive a little deeper into the fundamental problem here, solutions of other libraries, and what we can do in our situation.

== What's the problem here?

> In my opinion, the fundamental problem is that Hibernate Reactive doesn't return 'regular' ``Uni``'s, but ``Uni``'s with an additional instruction manual. The instruction manual (_don't evaluate them in parallel!_) doesn't show up in the type, so developers won't know about them until they read the manual. Worse, other libraries won't know about them either and violate their rules.

Reviewer: ```suggestion
In my opinion, the fundamental problem is that Hibernate Reactive doesn't return 'regular' ``Uni``s, but ``Uni``s with an additional instruction manual. The instruction manual (_don't evaluate them in parallel!_) doesn't show up in the type, so developers won't know about them until they read the manual. Worse, other libraries won't know about them either and violate their rules.
```

=== Another approach: Doobie

https://tpolecat.github.io/doobie/[Doobie] is a popular database access library for Scala, which has solved this by using a specific type `ConnectionIO`, which _does not allow parallel composition_, but only sequential composition. You _can_ convert from this type to a more general type (somewhat like `Uni`), but at that point you define the transaction boundary.

So the entire instruction manual of _you can not run database work on the same transaction in parallel_, which is similar to the constraint of Hibernate, is encoded in the type system. Beautiful.

> == Making your Uni's safe

Reviewer: ```suggestion
== Making your ``Uni``s safe
```

> So, the problem is that Hibernate ``Uni``'s should not be evaluated in parallel, and that libraries that consume the ``Uni``'s are not aware of this restriction.

Reviewer: ```suggestion
So, the problem is that Hibernate ``Uni``s should not be evaluated in parallel, and that libraries that consume the ``Uni``s are not aware of this restriction.
```

So naturally, we could ask: can we remove this constraint, and make our ``Uni``s safe for parallel execution?

The answer is that we can, and it's not terribly complicated.

> What we want to do is to shift the responsibility for sequential execution from outside the ``Uni``'s to _inside_ the ``Uni``'s. So that if two Hibernate Reactive ``Uni``s are run in parallel, one of them will wait with executing the actual Hibernate work until the other is done.

Reviewer: ```suggestion
What we want to do is to shift the responsibility for sequential execution from outside the ``Uni``s to _inside_ the ``Uni``s. So that if two Hibernate Reactive ``Uni``s are run in parallel, one of them will wait with executing the actual Hibernate work until the other is done.
```

> We will make a mutex for ``Uni``s that doesn't block threads, allowing multiple Uni's evaluated in parallel to sequence themselves, without blocking threads.

Reviewer: ```suggestion
We will make a mutex for ``Uni``s that doesn't block threads, allowing multiple ``Uni``s evaluated in parallel to sequence themselves, without blocking threads.
```

=== A Uni Mutex

> We want to create a semaphore, that we can use to protect a critical section:

Reviewer: ```suggestion
We want to create a semaphore that we can use to protect a critical section:
```

    private int permits;
    private final Queue<UniEmitter<Void>> queue;

    public UniSemaphoreImpl(int permits) {
        assert(permits > 0);
        this.permits = permits;
        queue = new LinkedBlockingDeque<>();
    }

    @Override
    public <T> Uni<T> protect(Uni<T> uni) {
        return acquire().replaceWith(uni).eventually(this::release);
    }

    private Uni<Void> release() {
        return Uni.createFrom().item(() -> {
            synchronized (this) {
                UniEmitter<Void> next = queue.poll();
                if (next == null) {
                    permits++;
                } else {
                    next.complete(null);
                }
                return null;
            }
        });
    }

    private Uni<Void> acquire() {
        return Uni.createFrom().deferred(() -> {
            synchronized (this) {
                if (permits >= 1) {
                    permits--;
                    return Uni.createFrom().voidItem();
                } else {
                    return Uni.createFrom().emitter(emitter -> queue.add((UniEmitter<Void>) emitter));
                }
            }
        });
    }
}

----

> The `protect` method will wrap the uni with work (typically the `Uni` doing Hibernate reactive stuff), between an `acquire` and a `release`.

Reviewer: ```suggestion
The `protect` method will wrap the ``Uni`` with work (typically the `Uni` doing Hibernate reactive stuff), between an `acquire` and a `release`.
```

I'm a bit confused with using one or two single quotes around **Uni**

`acquire` will see if there are permits, if so it'll take one and perform the work. If not, it will schedule the work in a queue and immediately return a `Uni`. This way, it doesn't block the thread if there's no permit available.

`release` will run when the work completes or when it fails and will either start queued work, or return the permit.

Now we can convert 'unsafe' ``Uni``s to 'safe' ``Uni``s, that a user can safely run in parallel:

[source, java]
----
Uni<String> unsafe1 = repo.getFoo();
Uni<String> unsafe2 = repo.getBar();

Uni.join().all(unsafe1, unsafe2).andFailFast(); // This will cause Hibernate exceptions

// Now make them 'safe':
UniSemaphore mutex = new UniSemaphoreImpl(1);
Uni<String> safe1 = mutex.protect(() -> unsafe1);
Uni<String> safe2 = mutex.protect(() -> unsafe2);

Uni.join().all(safe1, safe2).andFailFast(); // This will work properly
----

=== Alternative approach

> An alternative approach to executing ``Uni``'s in parallel, is making sure each ``Uni`` runs on their own Vertx context, with each their own Hibernate Reactive Session:

Reviewer: ```suggestion
An alternative approach to executing ``Uni``s in parallel, is making sure each ``Uni`` runs on their own Vertx context, with each their own Hibernate Reactive Session:
```
